components/Banner/examples/supplemental/MDCBannerViewLayout.m

```diff
@@ -0,0 +1,269 @@
+    UIButton *button = [self.dataSource bannerViewLayout:self buttonAtIndex:index];
+    [self.buttonFrames addObject:[NSValue valueWithCGRect:button.frame]];
+  }
+}
+
+// TODO: add cache support for views currently being used for performance
+- (void)updateLayoutStyleAndFrameSize {
+  // Single Line Style
+  CGFloat remainingWidth = self.sizeToFit.width;
+  remainingWidth -= (kLeadingPadding + kTrailingPadding);
+  remainingWidth -= [self buttonsWidthSum];
+  remainingWidth -= kHorizontalSpaceBetweenTextLabelAndButton;
+  if (self.iconImageView) {
+    remainingWidth -= self.iconImageView.frame.size.width;
+    remainingWidth -= kSpaceBetweenIconImageAndTextLabel;
+  }
+  CGSize textLabelSize = CGSizeZero;
+  if ([self isAbleToFitTextLabelWithWidthLimit:remainingWidth]) {
+    CGFloat singleLineStyleFrameHeight = kTopPaddingSmall + kBottomPadding;
+    NSMutableArray *singleLineViews = [[NSMutableArray alloc] init];
+    [singleLineViews addObject:self.textLabel];
+    if (self.iconImageView) {
+      [singleLineViews addObject:self.iconImageView];
+    }
+    singleLineStyleFrameHeight +=
+        MAX([self maximumHeightAmongViews:singleLineViews], [self maximumButtonHeight]);
+    self.style = MDCBannerViewLayoutSingleLineStyle;
+    textLabelSize = [self.textLabel sizeThatFits:CGSizeMake(remainingWidth, CGFLOAT_MAX)];
+    self.frameSize = CGSizeMake(self.sizeToFit.width, singleLineStyleFrameHeight);
+  } else {
+    CGFloat multiLineStyleFrameHeight = kTopPaddingLarge + kBottomPadding;
+    // Handle Buttons
+    if ([self buttonsWidthSum] + kLeadingPadding + kTrailingPadding > self.sizeToFit.width) {
+      self.style = MDCBannerViewLayoutMultiLineStackedButtonStyle;
+      multiLineStyleFrameHeight += [self buttonsHeightSum];
+    } else {
+      self.style = MDCBannerViewLayoutMultiLineAlignedButtonStyle;
+      multiLineStyleFrameHeight += [self maximumButtonHeight];
+    }
+    // Handle Image and Text
+    remainingWidth = self.sizeToFit.width - kLeadingPadding - kTrailingPadding;
+    if (self.iconImageView) {
+      remainingWidth -= (self.iconImageView.frame.size.width + kSpaceBetweenIconImageAndTextLabel);
+      textLabelSize = [self.textLabel sizeThatFits:CGSizeMake(remainingWidth, CGFLOAT_MAX)];
+      multiLineStyleFrameHeight +=
+          MAX(textLabelSize.height + kVerticalSpaceBetweenButtonAndTextLabelLarge,
+              76.0f);  // Hard coded value here (Image side + padding)
+    } else {
+      multiLineStyleFrameHeight += kVerticalSpaceBetweenButtonAndTextLabelSmall;
+      textLabelSize = [self.textLabel sizeThatFits:CGSizeMake(remainingWidth, CGFLOAT_MAX)];
+      multiLineStyleFrameHeight += textLabelSize.height;
+    }
+    self.frameSize = CGSizeMake(self.sizeToFit.width, multiLineStyleFrameHeight);
+  }
+  self.textLabel.frame = CGRectMake(0.0f, 0.0f, remainingWidth, textLabelSize.height);
+}
+
+- (void)updateLayoutWithStyle:(MDCBannerViewLayoutMode)layoutMode andFrameSize:(CGSize)frameSize {
```

Comment:
If we're planning to make Banner available (ie: "ready") after we drop ios 8.0 support, than you may want to consider using stack views instead of manually laying out the buttons.

Reply:
Mark this one for discussion.

Reply:
Actionable Thing: check UIStackView